editor/src/clj/editor/pose.clj

```diff
@@ -0,0 +1,255 @@
+
+(defn scale-pose
+  ^Pose [^double x ^double y ^double z]
+  (if (and (== 1.0 x) (== 1.0 y) (== 1.0 z))
+    default
+    (->Pose default-translation default-rotation (vector-of :double x y z))))
+
+(defn make
+  ^Pose [translation rotation scale]
+  {:pre [(or (nil? translation) (s/valid? ::translation translation))
+         (or (nil? rotation) (s/valid? ::rotation rotation))
+         (or (nil? scale) (s/valid? ::scale scale))]}
+  (let [translated (significant-translation? translation)
+        rotated (significant-rotation? rotation)
+        scaled (significant-scale? scale)]
+    (if (or translated rotated scaled)
+      (->Pose (if translated translation default-translation)
+              (if rotated rotation default-rotation)
+              (if scaled scale default-scale))
+      default)))
+
+(defn from-translation
+  ^Pose [translation]
+  {:pre [(or (nil? translation) (s/valid? ::translation translation))]}
+  (if translation
+    (apply translation-pose translation)
+    default))
+
+(defn from-rotation
+  ^Pose [rotation]
+  {:pre [(or (nil? rotation) (s/valid? ::rotation rotation))]}
+  (if rotation
+    (apply rotation-pose rotation)
+    default))
+
+(defn from-euler-rotation
+  ^Pose [euler-rotation]
+  {:pre [(or (nil? euler-rotation) (s/valid? ::euler-rotation euler-rotation))]}
+  (if euler-rotation
+    (apply euler-rotation-pose euler-rotation)
+    default))
+
+(defn from-scale
+  ^Pose [scale]
+  {:pre [(or (nil? scale) (s/valid? ::scale scale))]}
+  (if scale
+    (apply scale-pose scale)
+    default))
+
+(defn pre-multiply
+  ^Pose [^Pose child-pose ^Pose parent-pose]
+  {:pre [(pose? child-pose)
+         (pose? parent-pose)]}
+  (let [child-translation (.-translation child-pose)
+        child-rotation (.-rotation child-pose)
+        child-scale (.-scale child-pose)
+        parent-translation (.-translation parent-pose)
+        parent-rotation (.-rotation parent-pose)
+        parent-scale (.-scale parent-pose)
+        parent-translated (significant-translation? parent-translation)
+        parent-rotated (significant-rotation? parent-rotation)
+        parent-scaled (significant-scale? parent-scale)]
+    (cond-> child-pose
+
+            parent-scaled
+            (assoc :scale (multiply-vector child-scale parent-scale))
+
+            parent-rotated
+            (assoc :rotation (rotate-quaternion child-rotation parent-rotation))
+
+            (or parent-scaled parent-rotated parent-translated)
+            (assoc :translation
+                   (cond-> child-translation
+                           parent-scaled (multiply-vector parent-scale)
+                           parent-rotated (rotate-vector parent-rotation)
+                           parent-translated (add-vector parent-translation))))))
+
+(defn to-map [^Pose pose translation-key rotation-key scale-key]
+  {:pre [(pose? pose)]}
+  {translation-key (.-translation pose)
+   rotation-key (.-rotation pose)
+   scale-key (.-scale pose)})
```

Comment:
Maybe we should instead change the code that uses `to-map` to expect `:translation`/`:rotation`/`:scale` keys? I worry that this fn enables us to introduce more inconsistencies in pose-related code. The pose record is already a map...

Or maybe the pose should use names `:position`/`:rotation`/`:scale3`?

Reply:
These inconsistencies are unfortunately placed upon us by the requirements of the existing file formats. Scale is named `:scale3` in collections because we used to have a uniform `:scale` that was a single `float`, but for game objects the non-uniform (i.e vector triplet) scale field is simply named `:scale`. Likewise, the translation field for both game objects and collections are named `:position`, even though it is in fact a translation along the (possibly rotated and scaled) axes of its parent. It is correctly (in my opinion) named `:translation` in the protobuf `Transform` message, which our Pose type use-case mirrors.

I would have preferred to call the `Pose` record `Transform` in the editor as well. Unfortunately the term `transform` is used to refer to `Matrix4` values everywhere in the editor, and both the Spine and Rive extensions already rely on the `transform` output of the base `SceneNode` they subclass to produce a `Matrix4`. Maybe we can clean this up in the future, but we'd have to do it onto the `dev` since it is a breaking change requiring projects to update their library dependencies.

For now, I prefer to do the map conversion like this. In fact nothing outside of the unit tests make assumptions about what the record fields are called, or even that the Pose is represented as a record. If we ever get around to migrating away from `javax.vecmath`, we might want to represent it differently.